lib/booking_system.rb

```diff
@@ -0,0 +1,118 @@
+module Hotel
+  class BookingSystem
+    attr_accessor :rooms, :reservations, :all_ids, :blocks
+
+    def initialize
+      @rooms = load_rooms
+      @reservations = []
+      @all_ids = []
+      @blocks = []
+    end
+
+    def load_rooms
+      rooms = []
+      i = 1
+
+      20.times do
+        rooms << i
+        i += 1
+      end
+      return rooms
+    end
+
+    def make_reservation(start_date, end_date)
+      reservation = Hotel::Reservation.new(reservation_id: generate_id, room: assign_available_room(start_date, end_date), start_date: start_date, end_date: end_date, price_per_night: 200)
+      @reservations << reservation
+      # binding.pry
+    end
+
+    def make_block(start_date, end_date, number_of_rooms)
+      if number_of_rooms > 5
+        raise ArgumentError, "Cannot reserve more than 5 rooms"
+      end
+
+      id = generate_id
+
+      block = Hotel::Block.new(block_id: id, number_of_rooms: number_of_rooms, start_date: start_date, end_date: end_date, price_per_night: 150)
+      @blocks << block
+
+      number_of_rooms.times do
+        block_reservation = Hotel::BlockReservation.new(block_id: id,  reservation_id: nil, room: assign_available_room(start_date, end_date), start_date: start_date, end_date: end_date, price_per_night: 150)
+        @reservations << block_reservation
+      end
+    end
+
+    def make_block_reservation(block_id)
+      block_reservation = find_empty_block_reservation(block_id)
+      block_reservation.reservation_id = generate_id
+      return block_reservation
+    end
+
+    def assign_available_room(start_date, end_date)
+      booked_rooms = []
+      available_rooms = []
+      @reservations.each do |reservation|
+        if
+          reservation.start_date < end_date && start_date < reservation.end_date
+          booked_rooms << reservation.room
+        end
+      end
+      all_rooms = load_rooms
+      available_rooms = all_rooms - booked_rooms
+      if available_rooms[0] == nil
+        raise ArgumentError, "No rooms available"
+      end
+      return available_rooms[0]
+    end
+
+    def search_reservations(start_date, end_date)
+      reservations_within_date = []
+      @reservations.each do |reservation|
+        if
+          reservation.start_date < end_date && start_date < reservation.end_date
```

Comment:
This and the above method could be streamlined a little using the `select` enumerable:

```ruby
return @reservations.select do |reservation|
  reservation.start_date < end_date && start_date < reservation.end_date
end
```